core/unix/module_elf.c

```diff
@@ -1133,11 +1132,21 @@ module_lookup_symbol(ELF_SYM_TYPE *sym, os_privmod_data_t *pd)
      * FIXME: i#461 We do not tell weak/global, but return on the first we see.
      */
     ASSERT_OWN_RECURSIVE_LOCK(true, &privload_lock);
-    mod = privload_first_module();
     /* FIXME i#3850: Symbols are currently looked up following the dependency chain
      * depth-first instead of breadth-first.
      */
-    while (mod != NULL) {
+    for (privmod_t *mod = privload_first_module(); mod != NULL;
+         mod = privload_next_module(mod)) {
+        /* Skip other client modules at this point because some will not be
+         * initialised and clients should be leaves of the dependency tree and
+         * not provide symbols for other modules. Skipping just the uninitialised
+         * client modules should also work but might introduce an element of
+         * unpredictability if we are unsure in what order modules will be
+         * initialised. Skipping all uninitialised modules should also work but
+         * might hide a more serious problem. See i#4501.
+         */
+        if (mod->is_client)
```

Comment:
At least on Linux, it looks as though `is_client` is true only for the top-level client libraries, so the comment is wrong. I tried replacing the `mod->is_client = true` in `loader.c` with a syscall exit_group and nothing broke so it looks as though that bit of code in `loader.c` is indeed "stale", as the comment states, though I can't explain why the condition there is never true.

Reply:
It does seem an audit of `is_client` uses is needed to determine whether it should include the libraries.  One potential short-term solution to avoid any murkiness is to add a new flag `is_top_level_client` or somesuch name that is only set for these early-mapped-but-not-initialized client libs with a XXX comment asking whether it could replace `is_client` -- which as you said has some complexity in figuring out the intent and consequences of possibly changing behavior.  In my memory I thought the libraries were also treated as clients for interfaces like dr_memory_is_in_client() -- how does that relate to this `is_client`.  Seems some cleanup/clarification is needed, which doesn't have to block this PR if we add a new field.

Reply:
Done.